### PR TITLE
erasure code: shec performance optimization

### PR DESCRIPTION
**Pull request by t-miyamae**
Optimize shec's cpu utilization by caching decoding matrix and using intel sse4

This pull request is a split of #4083.
Hope these commits will be backported to a hammer point release in the near future.

Signed-off-by: Takeshi Miyamae miyamae.takeshi@jp.fujitsu.com

### CODE REVIEW

**Comment by liewegas** [comment]
Looks like this needs a rebase at least...

**Comment by t-miyamae** [comment]
We've rebased our commits (solved conflicts).

**Comment by dachary** [comment]
I'll review that in the next few days, thanks for your patience.

File: src/erasure-code/shec/Makefile.am [hunk 1]
```diff
@@ -23,33 +41,123 @@ libec_shec_la_SOURCES = \
...
+endif
+
+libec_shec_la_SOURCES = \
+	erasure-code/shec/ErasureCodePluginSelectShec.cc
+
+libec_shec_la_CFLAGS = ${AM_CFLAGS}
+libec_shec_la_CXXFLAGS= ${AM_CXXFLAGS}
 libec_shec_la_LIBADD = $(LIBCRUSH) $(PTHREAD_LIBS) $(EXTRALIBS)
 #libec_shec_la_LIBADD = $(PTHREAD_LIBS) $(EXTRALIBS)
 #libec_shec_la_LDFLAGS = ${AM_LDFLAGS} -version-info 1:0:0 -export-symbols-regex '.*__erasure_code_.*'
```
Comment (dachary) [inline src/erasure-code/shec/Makefile.am]: these commented lines can be removed

**Comment by dachary** [comment]
The title **shec performance optimization by intel sse4** should be changed to reflect the fact that it also contains optimizations for ARM processors. Something like **shec performance optimization with SIMD instructions** ?

**Comment by dachary** [comment]
Could you add tests similar to TestErasureCodePluginJerasure.cc to verify that the plugin selection logic actually works ? I'm not worried, from what I see it should work just fine. But it's useful to protect against regressions.

**Comment by dachary** [comment]
The commit "erasure code: shec performance optimization by decoding cache" contains multiple changes which makes it complicated to review. Could you create individual commits for the following:

* code is moved from shec.cc into double ErasureCodeShec::shec_calc_recovery_efficiency1 etc.
* ErasureCodeShecTableCache.{cc,h} updated with changes picked from ErasureCodeIsaTableCache.{cc,h}
* ErasureCodeShecTableCache::putDecodingTableToCache and getDecodingTableFromCache

**Comment by dachary** [comment]
From what I understand of getDecodingTableFromCache / putDecodingTableFromCache, you need to have more parameters than what ISA has ( matrixtype, k, m) and the DecodingCacheParameter seems a good idea to encapsulate them. Instead of having

<pre>
+  bool getDecodingTableFromCache(int* matrix,
+				 int* dm_ids,
+				 int* minimum,
+                                 int technique,
+                                 int k,
+                                 int m,
+				 int c,
+				 int w,
+				 int* erased,
+				 int* avails);
</pre>
which is the equivalent of the ISA table cache function
<pre>
  bool getDecodingTableFromCache(std::string &signature,
                                 unsigned char* &table,
                                 int matrixtype,
                                 int k,
                                 int m);
</pre>
you could have something like
<pre>
  bool getDecodingTableFromCache(std::string &signature,
                                 unsigned char* &table,
                                 DecodingCacheParameter& parameter);
</pre>
In other word, make it so getDecodingTableFromCache and putDecodingTableFromCache do not have a hard coded list of input/output arguments but an input / output class. 

What do you think ?

**Comment by dachary** [comment]
@t-miyamae do you have benchmarks that show how much is saved by caching the decoding matrix ? I've done some preliminary tests on jerasure and it looks like it does not improve because the matrix inversion is 1/1000 of the computational cost of recovering two lost chunks. Am I missing something ?

**Comment by t-miyamae** [comment]
Thank you for your restarting review of our pull request, Loic-san! We are handling your comments now.
By the way, as for the effect of decoding cache, we struggled and finally gained 50% improvement of minimum_to_decode's processing time. We can tell you the following facts.
(1) Either reducing index access frequency of decode_tbls_map or reducing the size of signature took effect.
(2) Our decoding cache may have taken effect because SHEC's processing in minimum_to_decode was originally heavier.

**Comment by dachary** [comment]
@t-miyamae I missed the fact that the minimum_to_decode function calls shec_make_decoding_matrix. 

https://github.com/ceph/ceph/blob/master/src/erasure-code/shec/ErasureCodeShec.cc#L119

The minimum_to_decode function is called  on every read operation, regardless of whether it requires decoding or not. https://github.com/ceph/ceph/blob/master/src/osd/ECBackend.cc#L1360

Don't you think it would be more efficient to pre-compute all minimum_to_decode answers into a table when the object is initialized ?

**Comment by t-miyamae** [comment]
@dachary I totally agree with you. Calculating all minimum_to_decode answers and decoding matrices at init() function must be more efficient. If you implement the framework, we will follow you.

**Comment by dachary** [comment]
@t-miyamae cool :-) Could you create a pull request with just erasure code: shec performance optimization by intel sse4 ? This commit is good to merge (modulo the remarks I made) and we don't need to wait for the caching logic to do so.

**Comment by t-miyamae** [comment]
@dachary Sure. If the target of the new caching logic is v0.94.2, we do not have to commit old one.

**Comment by t-miyamae** [comment]
Sorry, all the commits requested today were just mistakes and are not needed to be reviewed. We will open another pull request from new branch wip-mshec-sse4opt2 (not from wip-mshec-sse4opt).

**Comment by dachary** [comment]
@t-miyamae ok, thanks for the heads up :-) Fee free to close this pull request, another one can be open later.

**Comment by t-miyamae** [comment]
@dachary This pull request is already continued by #4689.

**Comment by athanatos** [comment]
@dachary

This needs tests added to ceph-qa-suite.

**Comment by dachary** [comment]
@athanatos ack. @t-miyamae I know you've worked to prepare teuthology tests, how close are you to completion ?

**Comment by t-miyamae** [comment]
@dachary We found an issue when running teuthology and we are trying to fix it. I'm not sure when we can settle it but we hope in a few weeks.

**Comment by dachary** [comment]
@t-miyamae could you please close this pull request and re-open before pushing a new version of the patch ?

**Comment by t-miyamae** [comment]
@dachary Thank you for merging #4976. We will do next pull request of SIMD and decoding cache optimization tomorrow. As for decoding cache, we will commit old implementation because we are sure that I/F changes discussed above has not finished yet. Would you tell us if our understanding is something wrong?

**Pull request closed without merging**